design-proposals/klip-15-new-api-and-client.md

```diff
@@ -0,0 +1,325 @@
+
+The request method will be a POST.
+
+Requests will be made to a specific URL, e.g. "/query-terminate" (this can be configurable)
+
+The body of the request is a JSON object UTF-8 encoded as text, containing the arguments for the
+operation (newlines have been added here for the sake of clarity but the real JSON must not contain newlines)
+
+````
+{
+"query-id": "xyz123", <----- the ID of the query to terminate
+}
+
+````
+ 
+
+#### Inserts
+
+The request method will be a POST.
+
+Requests will be made to a specific URL, e.g. "/insert-stream" (this can be configurable)
+
+The body of the request is a JSON object UTF-8 encoded as text, containing the arguments for the
+operation (newlines have been added for clarity, the real JSON must not contain newlines):
+
+````
+{
+"stream": "my-stream" <----- The name of the KSQL stream to insert into
+"acks": true          <----- If true then a stream of acks will be returned in the response
+}
+
+````
+
+Followed by zero or more JSON objects representing the values to insert:
+
+````
+{
+"col1" : "val1",
+"col2": 2.3,
+"col3", true
+}
+````
+Each JSON object will be separated by a new line.
+
+To terminate the insert stream the client should end the request.
+
+If acks are requested then the response will be written to the response when each row has been
+successfully committed to the underlying topic. Rows are committed in the order they are provided.
+Each ack in the response is just an empty JSON object, separated by newlines:
+
+````
+{}
+{}
+{}
+{}
+````
+
+#### Errors
+
+Apropriate status codes will be returned from failed requests. The response will also contain JSON
+with further information on the error:
+
+{
+    "error_code": <Error code>
+    "message": <Error Message>
+}
+
+#### Non streaming results
+
+The API is designed for efficiently streaming rows from client to server or from server to client.
+
+The amount of data that is streamed in any specific query or insert stream can be huge so we want to
+avoid any solution that buffers it all in memory at any one time or requires specialised parsers to
+parse.
+
+For this reason we do not provide query results (or accept streams of inserts) by default as a single
+JSON object. If we did so we would force users to find and use a streaming JSON parser in order to 
+parse the results as they arrive. If no parser is available on their platform they would be forced
+to parse the entire result set as a single JSON object in memory - this might not be feasible or
+desirable due to memory and latency constraints. Moreover, in the world of streaming it's quite common
+to want to pipe the stream from one place to another without looking into it. In that case it's very
+inefficient to deserialize the bytes as JSON simply to serialize them back to bytes to write them
+to the output stream. For these reasons the results, by default, are sent as a set of JSON arrays
```

Comment:
It's quite a common pattern in messaging (e.g. when implementing a router) to want to only inspect the headers of a message and not deserialise the body of the message for efficiency reasons. If the whole envelope was JSON you'd have to deserialise the whole thing. In the future we might want to introduce headers outside of the JSON body.

Reply:
I will however remove this sentence as we don't support headers (yet)

Reply:
That makes sense -- it not clear from the original write up though. It sounds like as if _different_ message are sent a JSON array, what only helps to get efficient access to a single message. (And does not help at all for the use case you describe.)

If we want to support headers, each message itself would need to be represented a s JSON array though, with explicit record/header "slot" for example. Is that what you tried to say?

Atm it seems that a single record is a single JSON with a simple tuple layout (not even the Kafka internal attribute-to-key/value mapping can be inferred -- what is desired I guess -- only the schema metadata would allow to infer the mapping -- but we want to hide the internal Kafka key/value model anyway I guess).

Just following up to make sure I understand correctly.